### ISSUE TITLE
Exposition of var@n syntax to the surface language

### ISSUE THREAD

**Issue by Profpatsch** [dhall-lang/dhall-lang#556]
@Profpatsch @ari-becker: The primary motivation for the `@n` language feature is to ensure that both the user and interpreter don't have to change the variable name if they wish to preserve the name.  This comes in handy in four ways:

* α-normalization

  @ari-becker: This is probably the use case you remembered:

  ```bash
  $ dhall normalize --alpha <<< 'λ(x : Natural) → λ(y : Bool) → x'
  λ(_ : Natural) → λ(_ : Bool) → _@1
  ```

  In α-normalization, all of the variable names *must* be `_`, so the only way to distinguish them is via the `@n` syntax.  In this scenario the `@n` suffix plays the role of a De Bruijn index.

* β-normalization

  You can also run into this `@`n feature even for ordinary normalization.  For example:

  ```bash
  $ dhall <<< 'λ(x : Natural) → let y = x in λ(x : Bool) → y'
  λ(x : Natural) → λ(x : Bool) → x@1
  ```

  Here the interpreter is trying to always preserve the variable name that the user selected (`x` in this case), but in order to do so the interpreter must have a way to avoid variable capture.  The `@n` feature allows the interpreter to avoid such capture while still preserving the name the user assigned.

  @ari-becker: This is an example of how even users who did not intend to use name shadowing can still produce expressions whose normal forms have shadowed variable names.

* User preference

  Some users (like @ari-becker) might have no intention of referring to shadowed variable names, but others users might prefer to do so, perhaps because they prefer to reuse the same variable name multiple times within a large expression.

  For example, Haskell programmers do this all the time, except adding numeric suffixes (i.e. `s0`, `s1`, `s2`, etc.) or multiple `'`s to the end of their variable names (i.e. `s`, `s'`, `s''`, etc.).  The `@n` feature gives them a more systematic way of doing so that has language support.

* Ease of implementation

  The `@n` feature greatly simplifies the implementation of the language because implementations never need to generate fresh unused variable names.  The β-normalization algorithm (taken from Pierce's "Types and Programming Languages" book) is rather elegant and leads to normal forms whose variable names match the original names as closely as possible, using the minimum De Bruijn index necessary to avoid conflicts.

  Or, to put it another way, you avoid the issue that other programming languages (like PureScript) have where you have ugly inferred types like:

  ```purescript
  main :: forall t38 t39 t42 t47.
    ( MonadRec t39
    , Parallel t42 t39
    , MonadEff
        ( console :: CONSOLE
        | t47
        )
        t39
    ) => t39 t38
  ```

  The reason PureScript has these large numeric suffixes is because it's using a less elegant algorithm for avoiding name conflicts.  Dhall avoids this issue through the use of Pierce's algorithm but that algorithm requires language support for De Bruijn indices (i.e. the `@n`) and because those indices might appear in normal forms (see the second point above) they need to be made available to the user, too, not just the interpreter.

I agree that this features is missing from the documentation, and I hope to remember to fix that as part of moving the Haskell tutorial to a language-independent tutorial.

_Originally posted by @Gabriel439 in https://github.com/dhall-lang/dhall-haskell/issues/860#issuecomment-494036787_

**Comment by Gabriel439** [comment]
@Profpatsch: The `var@n` syntax is already available to the user.  The only thing missing is to document the feature.

**Comment by Profpatsch** [comment]
I would argue that there are two parts to this feature, the internal and the external part.
I agree that using numbered variables in the normalization pipeline is a huge simplification. But I’m not so sure about exposing it to the users.

Let’s say a user wants to shadow variables, as described above:

```
let x = "foo"
-- …
let x = "bar"
-- …
in (\(x: Text) -> x@1) "baz"
===
"bar"
```

Now the user refactors the code a bit, by moving around various let-bindings:

```
let x = "bar"
-- …
let x = "foo"
-- …
in (\(x: Text) -> x@1) "baz"
===
"foo"
```

They could also restructure their function a bit:

```
```
let x = "foo"
-- …
let x = "bar"
-- …
-- the variable is not used
in (\(_: Text) -> x@1) "baz"
===
"foo"
```

Both cases leading to very hard to debug problems because of simple refactors.

I think those are pretty serious and not worth the convenience gained by exposing the variable shadowing back-references to the surface language.

**Comment by Gabriel439** [comment]
@Profpatsch: I don't believe disallowing the `@n` syntax for the user is even an option.

As I mentioned in the second point: *even if the user never uses the shadowing feature* they can author expressions whose normal forms contain shadowed variables:

```
$ dhall <<< 'λ(x : Natural) → let y = x in λ(x : Bool) → y'
λ(x : Natural) → λ(x : Bool) → x@1
```

If the normal form were not itself syntactically valid that would cause a large number of issues with the user experience and toolchain.

**Comment by Profpatsch** [comment]
I see. I’d like to see e.g. user tooling discourage the manual use of indexing nonetheless, encouraging the user to explicitely re-name variables at definition site instead.

**Comment by singpolyma** [comment]
> I’d like to see e.g. user tooling discourage the manual use of indexing nonetheless, encouraging the user to explicitely re-name variables at definition site instead.

Why? This is a matter of preference and style for the user.‎ Being able to access any variable or code in full indices is a very attractive feature of Dhall for me.

**Comment by Profpatsch** [comment]
Because of the reasons I listed above. It’s a very brittle feature.

**Comment by Gabriel439** [comment]
@Profpatsch: But the problem you described would still be present even without user-facing support for `var@n` syntax.  Consider this example:

```haskell
let x = 1

let x = 2

in  x
```

Reordering `let` bindings would break that example, too, and that doesn't use `@n` at all

**Comment by Profpatsch** [comment]
That’s true, but also because compilers usually warn about variable shadowing.

**Comment by Nadrieril** [comment]
Now that I think about it, I wouldn't mind if normalizing renamed variables to avoid shadowing. E.g. `λ(x : Natural) → let y = x in λ(x : Bool) → y` would normalize to `λ(x : Natural) → λ(x_1 : Bool) → x` and alpha-normal forms would use variables named `_0`, `_1`, ... This would make normal forms vastly more readable at least.
This could make the spec more complicated but I'm not even sure since we might be able to get rid of `shift`.

**Comment by Nadrieril** [comment]
It would also help alleviate the problem of the alpha-normalization of cached imports since alpha-normalized values would be much more readable

**Comment by Gabriel439** [comment]
@Nadrieril: I personally don't think we should overload the `Text` value of a variable name to store the index.  That seems like a more weakly-typed way of doing it

**Comment by Nadrieril** [comment]
If that's the issue, we could easily say this is a new syntax and parse `x_1` as `("x", 1)` internally. Ah, in fact we could reuse the `@` syntax. So maybe I'm out of scope for this issue ^^'. Turns out I'm mainly advocating for reordering the indices to make normal forms more legible.

**Comment by ari-becker** [comment]
I just want to say, perhaps ironically as I was the user who perhaps might not ever have had a need for shadowed variable names, that I did find a use-case for shadowed variable names, as it allowed me to retain alphabetized order of let statements in a case where referencing the unshadowed variable name itself would have referenced the wrong scope. Having to rename variables to avoid the shadow syntax would, in my current level of understanding of the language, make my code less clean. So I "get" the feature now.

I didn't author this issue so it's not mine to close, but if I had authored this issue, I'd close it.

**Comment by Profpatsch** [comment]
I think the point about variable shadowing (and compilers warning about it) is still valid, because in >90% of cases it’s not what users want to do, but hints at a bug.
